src/main/java/org/traccar/protocol/ZrProtocolDecoder.java

```diff
@@ -0,0 +1,323 @@
+        // body
+        buffer.writeBytes(body);
+        // check sum
+        buffer.writeByte(12);
+        // msg tail
+        buffer.writeBytes(Unpooled.wrappedBuffer(ByteBufUtil.decodeHexDump(FRAME_TAIL)));
+
+        // Correct message length
+        buffer.setShort(2, buffer.readableBytes() - 6);
+        buffer.setByte(buffer.writerIndex() - 3, bcc(buffer));
+
+        return buffer;
+    }
+
+    private static void sendGeneralResponse(Channel channel, SocketAddress remoteAddress, ByteBuf id, int frameType, short editionNum, byte encryptionType, Integer packageNo) {
+        ByteBuf response = Unpooled.buffer();
+        add2391(response);
+        add23A0(response, frameType, packageNo);
+        add23A3(response);
+        if (channel != null) {
+            channel.writeAndFlush(new NetworkMessage(formatMessage(FRAME_TYPE_SER_GEN_ACK, id, editionNum, encryptionType, packageNo, response), remoteAddress));
+        }
+    }
+
+    private static void sendLbsTransferResponse(Channel channel, SocketAddress remoteAddress, ByteBuf id, int frameType, short editionNum, byte encryptionType, Integer packageNo) {
+        ByteBuf response = Unpooled.buffer();
+        add2391(response);
+        add2094(response);
+        add23A3(response);
+        if (channel != null) {
+            channel.writeAndFlush(new NetworkMessage(formatMessage(FRAME_TYPE_LBS_TRANSFER_LONG_LAT_ACK, id, editionNum, encryptionType, packageNo, response), remoteAddress));
+        }
+    }
+
+    private static void sendTimeSynResponse(Channel channel, SocketAddress remoteAddress, DeviceSession deviceSession, ByteBuf id, int frameType, short editionNum, byte encryptionType, Integer packageNo) {
+        ByteBuf response = Unpooled.buffer();
+        add2391(response);
+        add2510(response, deviceSession);
+        if (channel != null) {
+            channel.writeAndFlush(new NetworkMessage(formatMessage(FRAME_TYPE_TIME_SYN_ACK, id, editionNum, encryptionType, packageNo, response), remoteAddress));
+        }
+    }
+
+    public static void add2391(ByteBuf bodyBuf) {
+        bodyBuf.writeShort(0x2391);
+        bodyBuf.writeShort(3);
+        bodyBuf.writeBytes(ByteBufUtil.decodeHexDump("123456"));
+    }
+
+    private static void add23A0(ByteBuf bodyBuf, int frameType, Integer packageNo) {
+        bodyBuf.writeShort(0x23a0);
+        bodyBuf.writeShort(7); // len
+
+        bodyBuf.writeShort(packageNo);
+        bodyBuf.writeByte(0);
+        bodyBuf.writeShort(frameType);
+        bodyBuf.writeShort(0);
+    }
+
+    private static void add23A3(ByteBuf bodyBuf) {
+        bodyBuf.writeShort(0x23a3);
+        bodyBuf.writeShort(4);
+        long time = System.currentTimeMillis() / 1000;
+        bodyBuf.writeInt((int) time);
+    }
+
+    private static void add2094(ByteBuf bodyBuf) {
+        bodyBuf.writeShort(0x2094);
+        bodyBuf.writeShort(8);
+        bodyBuf.writeInt(0); // longitude
+        bodyBuf.writeInt(0); // latitude
```

Comment:
There make no sense to me at all. You have to explain.

Reply:
This is a protocol requirement, data provided by the platform.

Reply:
All those should have proper names.